### PR TITLE
[TradingMode] support dsl calls

### PR DESCRIPTION
**Pull request by GuillaumeDSM**

### CODE REVIEW

File: additional_tests/exchanges_tests/abstract_authenticated_exchange_tester.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ class AbstractAuthenticatedExchangeTester:
     # Add method call to subclasses to be able to run them independently
 
     async def test_get_portfolio(self):
-        # encoded_a = _get_encoded_value("") # tool to get encoded values
+        # encoded_a = _get_encoded_value("a") # tool to get encoded values
```
Comment (Herklos) [inline additional_tests/exchanges_tests/abstract_authenticated_exchange_tester.py]: 😄

File: packages/commons/octobot_commons/dsl_interpreter/operators/re_callable_operator_mixin.py [hunk 2]
```diff
@@ -56,6 +59,24 @@ def get_next_call_time(self) -> typing.Optional[float]:
             return last_execution_time + waiting_time
         return None
 
+    @staticmethod
+    def get_script_override(result: typing.Any) -> typing.Optional[str]:
```
Comment (Herklos) [inline packages/commons/octobot_commons/dsl_interpreter/operators/re_callable_operator_mixin.py:63]: 👍

File: packages/commons/octobot_commons/asyncio_tools.py [hunk 3]
```diff
@@ -128,8 +128,9 @@ def logged_waiter(self, name: str, sleep_time: float = 30) -> typing.Generator[N
     async def _waiter() -> None:
         t0 = time.time()
         try:
-            await asyncio.sleep(sleep_time)
-            self.logger.info(f"{name} is still processing [{time.time() - t0:.2f} seconds] ...")
+            while True:
+                await asyncio.sleep(sleep_time)
+                self.logger.info(f"{name} is still processing [{time.time() - t0:.2f} seconds] ...")
```
Comment (Herklos) [inline packages/commons/octobot_commons/asyncio_tools.py:133]: Why `while True?` is it an asyncio task to be canceled?

File: packages/copy/octobot_copy/copiers/account_copier.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    async def _resync_if_mirrored_open_order_grace_period_elapsed(self) -> None:
+        copy_settings = self._copy_settings
+        grace_seconds = copy_settings.mirrored_orphan_cancel_grace_seconds
+        grace_started_at = copy_settings.mirrored_orphan_grace_started_at
+        if (
+            grace_seconds > 0
+            and grace_started_at is not None
+            and grace_started_at > 0
+            and (time.time() - grace_started_at) >= grace_seconds
+        ):
```
Comment (Herklos) [inline packages/copy/octobot_copy/copiers/account_copier.py:96]: 👍

File: packages/copy/octobot_copy/rebalancing/planner/base_rebalance_actions_planner.py [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+                )
+                target_ratio = round(
+                    rebalance_details[rebalancer_enums.RebalanceDetails.ADD.value].get(
+                        target,
+                        rebalance_details[rebalancer_enums.RebalanceDetails.BUY_MORE.value].get(
+                            target, trading_constants.ZERO
+                        )
+                    ) * trading_constants.ONE_HUNDRED,
+                    3
+                ) or "???"
```
Comment (Herklos) [inline packages/copy/octobot_copy/rebalancing/planner/base_rebalance_actions_planner.py]: What `"???"` used for ?

File: packages/copy/octobot_copy/rebalancing/planner/historical_configuration_rebalance_actions_planner.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                    f"{self.client.get_ideal_distribution(historical_config)}."
+                )
+                return historical_config
+        self.logger.info(
+            f"No suitable {self.client.client_name} config found: using latest distribution: "
+            f"{self.client.get_ideal_distribution(config)}."
+        )
+        return config
+
+    def _is_target_config_applied(self, config: dict, traded_bases: set[str]) -> bool:
```
Comment (Herklos) [inline packages/copy/octobot_copy/rebalancing/planner/historical_configuration_rebalance_actions_planner.py:162]: 👍

File: packages/flow/octobot_flow/entities/actions/actions_dag.py [hunk 7]
```diff
@@ -133,12 +159,27 @@ def _resolve_dsl_script(
...
+                value = _navigate_dict_path(value, dependency.result_path)
             resolved_dsl_script = octobot_commons.dsl_interpreter.apply_resolved_parameter_value(
-                resolved_dsl_script, dependency.parameter, dependency_action.result
+                resolved_dsl_script, dependency.parameter, value
+            )
+        reschedule_params = action.get_rescheduled_parameters()
+        for rescheduled_parameter, rescheduled_value in reschedule_params.items():
+            if script_override := octobot_commons.dsl_interpreter.ReCallingOperatorResult.get_script_override(rescheduled_value):
+                # the script override is the new DSL script to execute for this action call
+                resolved_dsl_script = script_override
```
Comment (Herklos) [inline packages/flow/octobot_flow/entities/actions/actions_dag.py:172]: 👍

File: packages/flow/octobot_flow/entities/automations/execution_details.py [hunk 8]
```diff
@@ -40,11 +40,18 @@ class DegradedStateDetails(octobot_commons.dataclasses.MinimizableDataclass):
     reason: typing.Optional[str] = None
 
 
+@dataclasses.dataclass
+class CopyDetails(octobot_commons.dataclasses.MinimizableDataclass):
+    # wall time.time() when the current mirrored open-order grace episode started
+    open_orders_grace_period_started_at: typing.Optional[float] = None
```
Comment (Herklos) [inline packages/flow/octobot_flow/entities/automations/execution_details.py:46]: Do we still need it? Isn't it a kind of state we wanted to avoid?

File: packages/flow/octobot_flow/parsers/actions_dag_parser.py [hunk 9]
```diff
@@ -450,6 +532,59 @@ def _create_init_action(
...
+            refs.append((dsl_key, dep_action_id, result_path, value))
+        return refs
+
+    def _inject_dependency_placeholders_in_dsl_script(
+        self, dsl_script: str, refs: list[tuple[str, str, tuple[str, ...], str]]
+    ) -> str:
+        """
+        Turn dependency:: references in the translated DSL into UNRESOLVED_PARAMETER placeholders
+        and use keyword form when the value was emitted as a positional argument.
+        """
```
Comment (Herklos) [inline packages/flow/octobot_flow/parsers/actions_dag_parser.py:566]: 👍

File: packages/flow/octobot_flow/repositories/exchange/ohlcv_repository.py [hunk 10]
```diff
@@ -1,13 +1,32 @@
...
     ) -> exchange_data_import.MarketDetails:
-        return await exchanges_test_tools.fetch_ohlcv(
-            self.exchange_manager, symbol, time_frame, limit
+        updater = typing.cast(
+            octobot_trading.exchange_data.OHLCVUpdater,
+            self.get_channel_updater(octobot_trading.constants.OHLCV_CHANNEL)
+        )
+        ohlcvs = await updater.fetch_ohlcv(
+            symbol, common_enums.TimeFrames(time_frame), limit, allow_cache=True, tickers_backup=tickers
         )
```
Comment (Herklos) [inline packages/flow/octobot_flow/repositories/exchange/ohlcv_repository.py:21]: 👌

File: packages/flow/tests/functionnal_tests/trading_modes_actions/simulator/test_grid_trading_mode_action.py [hunk 11]
```diff
@@ -0,0 +1,719 @@
...
+    current_time,
+    resolved_actions,
+    automation_state_dict,
+    set_init_action_run_mode,
+    copy_exchange_account_action,
+    d_order_price,
+)
+
+import tentacles.Trading.Mode.grid_trading_mode.grid_trading as grid_trading
+
```
Comment (Herklos) [inline packages/flow/tests/functionnal_tests/trading_modes_actions/simulator/test_grid_trading_mode_action.py:34]: Great tests!

File: packages/node/tests/scheduler/test_tasks_recovery.py [hunk 12]
```diff
@@ -105,15 +109,32 @@ async def sleeper_workflow(identifier: float) -> float:
...
-            assert sorted(completed_workflows) == list(range(WF_TO_CREATE*2))
+            assert sorted(results_part2) == list(range(WF_TO_CREATE*2))
+            success_wfs = await octobot_node.scheduler.SCHEDULER.INSTANCE.list_workflows_async(
+                status=[dbos.WorkflowStatusString.SUCCESS.value]
+            )
+            assert len(success_wfs) == WF_TO_CREATE * 3
+            no_pending = await octobot_node.scheduler.SCHEDULER.INSTANCE.list_workflows_async(
+                status=[dbos.WorkflowStatusString.ENQUEUED.value, dbos.WorkflowStatusString.PENDING.value]
+            )
+            assert no_pending == []
```
Comment (Herklos) [inline packages/node/tests/scheduler/test_tasks_recovery.py:137]: 👍

File: packages/node/requirements.txt [hunk 13]
```diff
@@ -2,4 +2,4 @@ fastapi[standard]==0.135.1
 passlib[bcrypt]==1.7.4
 pydantic
 
-dbos==2.15.0
+dbos==2.17.0
```
Comment (Herklos) [inline packages/node/requirements.txt:5]: 🎉

File: packages/tentacles/Meta/DSL_operators/exchange_operators/exchange_personal_data_operators/copy_exchange_account_operators.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+) -> list[type[dsl_interpreter.PreComputingCallOperator]]:
+    class _CopyExchangeAccountOperator(dsl_interpreter.PreComputingCallOperator, dsl_interpreter.ReCallableOperatorMixin):
+        DESCRIPTION = (
+            "Rebalances the copier exchange toward the reference account allocation. "
+            "reference_account is JSON for octobot_copy.entities.Account (portfolio content, orders, positions). "
+            "account_copy_settings is optional JSON for AccountCopySettings; "
+            "reference market comes from the copier portfolio."
+        )
+        EXAMPLE = (
+            r"""copy_exchange_account(reference_account='{"content":{"BTC":{"available":"0.01","total":"0.01"}}}', """
```
Comment (Herklos) [inline packages/tentacles/Meta/DSL_operators/exchange_operators/exchange_personal_data_operators/copy_exchange_account_operators.py:52]: 👍

File: packages/tentacles/Trading/Mode/index_trading_mode/index_trading.py [hunk 15]
```diff
@@ -888,72 +590,75 @@ def init_user_inputs(self, inputs: dict) -> None:
...
+                symbol_util.merge_currencies(coin, reference_market)
+                for coin in previous_state.get(StateKeys.INDEXED_COINS.value, [])
+            ] if previous_state else []
+        )
+        return [trading_dsl.SymbolDependency(symbol=symbol) for symbol in list_util.deduplicate(symbols)]
+
+    def get_dsl_state(self) -> dict:
+        return {
+            StateKeys.INDEXED_COINS.value: self.indexed_coins,
+        }
```
Comment (Herklos) [inline packages/tentacles/Trading/Mode/index_trading_mode/index_trading.py:647]: 👍

File: packages/trading/octobot_trading/exchange_data/ticker/channel/ticker_updater.py [hunk 16]
```diff
@@ -101,6 +110,40 @@ async def trigger_ticker_update(self, symbol: str):
         self.logger.debug(f"Triggered ticker update for {symbol}")
         await self.fetch_and_push_pair(symbol)
 
+    async def fetch_all_tickers(self, symbols: typing.Optional[list[str]]) -> dict[str, dict]:
```
Comment (Herklos) [inline packages/trading/octobot_trading/exchange_data/ticker/channel/ticker_updater.py:113]: 👍

File: packages/trading/octobot_trading/modes/channel/abstract_mode_producer.py [hunk 17]
```diff
@@ -413,8 +420,25 @@ async def trading_mode_trigger(self, skip_health_check=False):
...
+        """
+        Called when a manual trigger is received. Behaves like set_final_eval by default.
+        """
+        return await self.set_final_eval(
+            matrix_id=matrix_id,
+            cryptocurrency=cryptocurrency,
+            symbol=symbol,
+            time_frame=time_frame,
+            trigger_source=trigger_source
+        )
```
Comment (Herklos) [inline packages/trading/octobot_trading/modes/channel/abstract_mode_producer.py:436]: 👍

File: packages/trading/octobot_trading/modes/abstract_trading_mode.py [hunk 18]
```diff
@@ -114,6 +120,11 @@ def __init__(self, config, exchange_manager):
         self.is_health_check_enabled: bool = False
         self._last_health_check_time: float = 0
 
+        # When True, the producer directly calls consumers instead of using channels.
+        # Awaiting the producer call completes after both producer and consumer work are done.
+        self.synchronous_execution: bool = False
+        self.previous_state: typing.Optional[dict] = None
```
Comment (Herklos) [inline packages/trading/octobot_trading/modes/abstract_trading_mode.py:126]: 👍

File: packages/trading/octobot_trading/personal_data/exchange_personal_data.py [hunk 19]
```diff
@@ -210,16 +231,14 @@ async def handle_portfolio_update_from_deposit(
         if should_notify:
             await self.handle_portfolio_update_notification(self.portfolio_manager.portfolio.portfolio)
 
+    @ignore_channel_notification_send_errors("Failed to send balance update notification")
```
Comment (Herklos) [inline packages/trading/octobot_trading/personal_data/exchange_personal_data.py:234]: 😮

**Review by Herklos — COMMENTED** [review]
Awesome work 💯 so many features !!

File: additional_tests/exchanges_tests/abstract_authenticated_exchange_tester.py (hunk 1 above)
Comment (GuillaumeDSM) [reply additional_tests/exchanges_tests/abstract_authenticated_exchange_tester.py]: 😮 i'll clear this

File: packages/commons/octobot_commons/asyncio_tools.py (hunk 3 above)
Comment (GuillaumeDSM) [reply packages/commons/octobot_commons/asyncio_tools.py:133]: it is, we want this task to log every 30s to inform that it's still in progress. it will be cancelled as soon as the wrapped task is done (it's a waiter)

File: packages/copy/octobot_copy/rebalancing/planner/base_rebalance_actions_planner.py (hunk 5 above)
Comment (GuillaumeDSM) [reply packages/copy/octobot_copy/rebalancing/planner/base_rebalance_actions_planner.py]: good question, it looks like an incorrect very old code (from https://github.com/Drakkar-Software/OctoBot/blob/dev/packages/tentacles/Trading/Mode/index_trading_mode/index_trading.py#L621) , i'll fix it

File: packages/flow/octobot_flow/entities/automations/execution_details.py (hunk 8 above)
Comment (GuillaumeDSM) [reply packages/flow/octobot_flow/entities/automations/execution_details.py:46]: we need this until the mechanism we discussed yesterday will be implemented (in the next PR)

File: packages/trading/octobot_trading/personal_data/exchange_personal_data.py (hunk 19 above)
Comment (GuillaumeDSM) [reply packages/trading/octobot_trading/personal_data/exchange_personal_data.py:234]: this is because of the re-use of the open source trading simulator and order state system by octobot flow: not all channels are started (it's normal), in the context, we don't want to log an error when a non-required channel push fails

File: packages/copy/octobot_copy/rebalancing/planner/base_rebalance_actions_planner.py (hunk 5 above)
Comment (GuillaumeDSM) [reply packages/copy/octobot_copy/rebalancing/planner/base_rebalance_actions_planner.py]: fix is up, it was just a log. It's now explicitely stated in the code

File: packages/commons/octobot_commons/asyncio_tools.py (hunk 3 above)
Comment (Herklos) [reply packages/commons/octobot_commons/asyncio_tools.py:133]: Are we sure all the previous usage (if they exists) of this function will have the same behavior with this loop?

Comment (GuillaumeDSM) [reply packages/commons/octobot_commons/asyncio_tools.py:133]: I don't understand what you mean, this is just a local task defined in the `async def _waiter()` function, there is no "previous usage" and nothing changes compared to before, the only thing is that the previous impl was missing a loop so the waiter would only log once, which was a mistake. The task is always cancelled when the context manager `logged_waiter` exits, this didn't change

**Review by Herklos — APPROVED** [review]
Great job 💯
